### PR TITLE
feat: add dynamic requirements for keep3rs on the job

### PR DESCRIPTION
**Pull request by lbertenasco**

### CODE REVIEW

File: contracts/keep3r/VaultKeep3r.sol [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+  // Collectable Dust
+  function sendDust(
+    address _to,
+    address _token,
+    uint256 _amount
+  ) external override onlyGovernor {
+    _sendDust(_to, _token, _amount);
+  }
+}
```
Comment (milkyklim) [inline contracts/keep3r/VaultKeep3r.sol:135]: 🔪

File: contracts/keep3r/VaultKeep3r.sol [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+  EnumerableSet.AddressSet internal availableVaults;
+
+  constructor(
+    address _keep3r,
+    address _bond,
+    uint256 _minBond,
+    uint256 _earned,
+    uint256 _age,
+    uint256 _earnCooldown
```
Comment (milkyklim) [inline contracts/keep3r/VaultKeep3r.sol:30]: Why do you need cooldown exactly?

File: contracts/keep3r/VaultKeep3r.sol [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    return (
+      calculateEarn(_vault) >= requiredEarn[_vault] &&
+      block.timestamp > lastEarnAt[_vault].add(earnCooldown)
+    );
+  }
+
+
+  // Keep3r actions
+  function earn(address _vault) external override paysKeeper {
+    require(workable(_vault), 'vault-keep3r::earn:not-workable');
```
Comment (milkyklim) [inline contracts/keep3r/VaultKeep3r.sol:100]: Do you need this line in `forceEarn`?

**Review by milkyklim — CHANGES_REQUESTED** [review]

File: contracts/keep3r/VaultKeep3r.sol (hunk 3 above)
Comment (lbertenasco) [reply contracts/keep3r/VaultKeep3r.sol:100]: Not really, forceEarn can only be called by the contract owner (previous strategist now).
No need to check since that action is not being rewarded with kp3r, so it cannot be abused.

File: interfaces/keep3r/IKeep3r.sol [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.8;
+interface IKeep3r {
```
[the remark this answers is not in the document]
Comment (lbertenasco) [reply interfaces/keep3r/IKeep3r.sol:3]: Yes, this is a bit confusing.
- keep3r (keep3r-network contract)
  - setKeep3r (sets the keep3r-network contract to be used)
- keeper (worker)
  - isKeeper (is msg.sender a keeper-worker on keep3r-network)

File: contracts/keep3r/VaultKeep3r.sol (hunk 2 above)
Comment (lbertenasco) [reply contracts/keep3r/VaultKeep3r.sol:30]: Explained on Telegram.
TLDR: for reducing the risk of getting the job credit drained.

File: scripts/vault/01-vault-keep3r-calculate-earn.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    try {
+      prompt.ask(async (answer) => {
+        if (answer) {
+          const [owner] = await ethers.getSigners();
+          // Setup deployer
+          await hre.network.provider.request({ method: "hardhat_impersonateAccount", params: [config.accounts.mainnet.deployer] });
+          const deployer = owner.provider.getUncheckedSigner(config.accounts.mainnet.deployer);
+          
+
+          // TODO REMOE
```
[the remark this answers is not in the document]
Comment (lbertenasco) [reply scripts/vault/01-vault-keep3r-calculate-earn.js:25]: Must remove before merging and after deploying :)

File: contracts/keep3r/Keep3rAbstract.sol [hunk 6]
```diff
@@ -33,4 +43,39 @@ contract Keep3r {
...
+    keep3r.workReceipt(msg.sender, _amount);
+  }
+  // Checks if caller is a valid keeper, handles payment amount in _credit after execution
+  modifier paysKeeperCredit(address _credit, uint256 _amount) {
+    _isKeeper();
+    _;
+    keep3r.receipt(_credit, msg.sender, _amount);
+  }
+  // Checks if caller is a valid keeper, handles payment amount in ETH after execution
+  modifier paysKeeperEth(uint256 _amount) {
```
Comment (dmolina79) [inline contracts/keep3r/Keep3rAbstract.sol:59]: don't see these modifiers used on the code, is that correct?

File: contracts/keep3r/VaultKeep3r.sol [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    require(requiredEarn[_vault] > 0, 'vault-keep3r::workable:vault-not-added');
+    return (
+      calculateEarn(_vault) >= requiredEarn[_vault] &&
+      block.timestamp > lastEarnAt[_vault].add(earnCooldown)
+    );
+  }
+
+
+  // Keep3r actions
+  function earn(address _vault) external override paysKeeper {
```
Comment (dmolina79) [inline contracts/keep3r/VaultKeep3r.sol:100]: small nitpick, prefer to use onlyKeeper and paysKeeper just pays keeper since otherwise it should really be named, `onlyKeeperAndPayKeeper` or something : /

**Review by dmolina79 — APPROVED** [review]
REviewed mainly the solidity changes, LGTM!

File: contracts/keep3r/Keep3rAbstract.sol (hunk 6 above)
Comment (lbertenasco) [reply contracts/keep3r/Keep3rAbstract.sol:59]: Yup, they are not being used, since we use the default payments for the current jobs.

File: contracts/keep3r/VaultKeep3r.sol (hunk 7 above)
Comment (lbertenasco) [reply contracts/keep3r/VaultKeep3r.sol:100]: I like this. I'll rework and submit.
